book/src/rent.md

```diff
@@ -0,0 +1,52 @@
+# Rent
+
+Accounts on Solana may have owner-controlled state (`Account::data`) that's separate 
+from the account's balance (`Account::lamports`).  Since validators on the network need 
+to maintain a working copy of this state in memory, the network charges a time-and-space 
+based fee for this resource consumption, also known as Rent.
+
+## Two-tiered rent regime
+
+Accounts which maintain a minimum balance equivalent to 2 years of rent are exempt from
+rent.  Accounts whose balance falls below this threshold are charged rent 
+at a rate specified in genesis, in lamports per kilobyte-year.
+
+## Collecting rent
+
+Rent is charged at the end of each slot on all accounts that were accessed during that block.  A field in the account `Account::rent_slot` is keeps track of the last time rent was paid by this account.
+
+If the account is in the exempt rent regime, `Account::rent_slot` is simply updated to the current slot.
+
+If the account not exempt, the difference between the current slot and `rent_slot` is used to calculate the number of years since the last rent payment, which is passed to `Rent::due()` to calculate the amount of rent owed by this account.  Any fractional lamports are truncated.  Rent is collected from the account and `Account::rent_slot` is updated to the current slot.
```

Comment:
Because rent is collected at the end of a block, it looks like I could bypass paying most rent by:
1) Create an account to hold my data with just enough lamports for the rent in which the block the account was created in.
2) When I want to transact using that account, my program accepts a second account as output.  The contents of the first account are copied into the output account before the first account is dropped.  That output account is also setup as in (1), with just enough lamports to get past Rent::due() for the block that created it.

Reply:
I don't understand this option.  All the allocated space outstanding at the end of the block is exposed to paying rent.  Allocating space will initialize rent_slot.

Reply:
Since rent isn't collected until the end of the block that contains a transaction that accessed my account, I can get one _free_ access to my account well after I'm delinquent.

Reply:
ok, got it.  a slot is very likely too granular a basis on which to collect rent, as the sigdigs end up going to zero

of the "others considered", do you have a favorite?

Reply:
I'm sorta liking epoch scans.  I suspect that we'll want to do this anyway to get purge dust system accounts too.   For example, the new `solana-wallet ping` command that I just added is basically a dust factory

Reply:
scan is going to be expensive :(.  i was thinking a "miner" can issue a transaction that collects rent for a mask of the accounts.

Reply:
we could also collect rent for an epoch in advance.  Account::rent_slot becomes Account::rent_epoch and means "the epoch this account is paid up for".  system_instruction::create_account collects initial rent.  accounts with no state have rent_epoch == 0.

Reply:
rent_epoch needs to be updated each time the account lamport balance changes right?  I think that works with CO accounts:
* if CO, rent_epoch is increased by the additional rent the new lamports pay for
* if CD, rent_epoch can be decreased safely

Reply:
CD and CO have washed away by the end of the slot, so we're actually modifying every account accessed in the slot that would have rent due as a result of the balance at the end of the slot

Reply:
CO and CD are washed away by the end of a slot (and the slot scan).  rent_epoch only needs to be pushed when rent is due and on `create_account()` where `!account.data.is_empty()`